### PR TITLE
Fix GraphEdit layering

### PR DESCRIPTION
**Pull request by Geometror**
This PR fixes/improves the ordering of comment nodes as children of GraphEdit as well as the position of the connections layer.
Split from #61414.

Detailed changes:
- the order of comment nodes is now calculated similar to a 2D BVH: when a comment node encloses other comment nodes, the index of the enclosing node is kept below
   - the order is updated when resizing a comment node
- the connections layer is now kept between the comment nodes and the normal graph (due to technical limitations it is no longer an internal node)

**Before:**
before

**This PR:**
after

### CODE REVIEW

**Comment by fire** [comment]
The nesting in GraphEdit::_top_layer_input is very deep. will review more.

**Comment by reduz** [comment]
This looks good to me, what is it missing?

**Comment by akien-mga** [comment]
> This looks good to me, what is it missing?

Mostly a contributor that feels confident enough with GraphEdit to approve the feature and implementation :)

**Comment by fire** [comment]
I'll take a look. Can it be rebased and ready to merge?

**Comment by Geometror** [comment]
@fire Rebased. 
Just for reference: For now(4.0) I think this will do, but since it is a bit hacky I would like to refactor the current GraphNode/GraphEdit system including the restructuring(#61414) and implementation of layers (or something similar) in the next few months after I opened a proper proposal. A problem might be compatibility breakage so that needs to be discussed.

**Comment by fire** [comment]
There's some conflicts.

**Review by fire — APPROVED** [review]

File: scene/gui/graph_edit.cpp [hunk 1]
```diff
@@ -387,6 +405,19 @@ void GraphEdit::_graph_node_slot_updated(int p_index, Node *p_gn) {
 	connections_layer->queue_redraw();
 }
 
+void GraphEdit::_graph_node_resized(Vector2 p_new_minsize, Node *p_gn) {
+	GraphNode *gn = Object::cast_to<GraphNode>(p_gn);
+	if (gn && gn->is_comment()) {
+		_update_comment_enclosed_nodes_list(gn, comment_enclosed_nodes);
+		_graph_node_raised(p_gn);
```
Comment (YuriSizov) [inline scene/gui/graph_edit.cpp:412]: Shouldn't it use the new method?

File: scene/gui/graph_edit.cpp [hunk 2]
```diff
@@ -1244,6 +1292,7 @@ void GraphEdit::gui_input(const Ref<InputEvent> &p_ev) {
 						gn->set_drag(false);
 						if (gn->is_comment()) {
 							_set_drag_comment_enclosed_nodes(gn, comment_enclosed_nodes, false);
+							_reorder_comment_nodes(gn);
```
Comment (YuriSizov) [inline scene/gui/graph_edit.cpp:1295]: I'm a bit confused by the order of operations. In the below snippet we update enclosed nodes after the reorder, and here we do before. It's a bit hard to follow the logic, so can you tell if this is intentional?

File: scene/gui/graph_edit.cpp [hunk 3]
```diff
@@ -399,7 +430,18 @@ void GraphEdit::add_child_notify(Node *p_child) {
 		gn->connect("selected", callable_mp(this, &GraphEdit::_graph_node_selected).bind(gn));
 		gn->connect("deselected", callable_mp(this, &GraphEdit::_graph_node_deselected).bind(gn));
 		gn->connect("slot_updated", callable_mp(this, &GraphEdit::_graph_node_slot_updated).bind(gn));
+
+		if (gn->is_comment()) {
+			background_nodes_separator_idx++;
```
Comment (YuriSizov) [inline scene/gui/graph_edit.cpp:435]: Do we actually need to track this? Can't we update connections layer's index along the comments? And in `_graph_node_raised` we can probably just refer to the connection layer's index.

I'm just concerned that adding a tracker like that would make the code too fragile.

**Comment by YuriSizov** [comment]
So, from some quick testing this doesn't seem to work completely as expected. For a new shader the comments are still in front of the connections:
image

But closing and reopening the file fixes it, even for the nodes added afterwards:
image

Am I doing something wrong?

**Comment by akien-mga** [comment]
Superseded by #67152.
